File: src/expressionparser/tokens/Function.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package expressionparser.tokens;

import expressionparser.EvaluateResult;
import expressionparser.IEvaluable;
import expressionparser.IToken;
import expressionparser.ITokenable;
import expressionparser.TokenData;
import java.util.ArrayList;

/**
 *
 * @author Lomztein
 */
public class Function implements IToken, ITokenable, IEvaluable {

    public Function(String name, int precedence, int argCount, Func function) {
        this.name = name;
        this.precedence = precedence;
        this.function = function;
        this.argCount = argCount;
    }

    private final String name;
    private final int precedence;
    private final Func function;
    private final int argCount;

    @Override
    public int getPrecedence() {
        return precedence;
    }

    private ArrayList<TokenData> getArguments(ArrayList<TokenData> from, int startIndex) {
        ArrayList<TokenData> arguments = new ArrayList<>();
        for (int i = startIndex + 1; i < from.size(); i++) {
            
            TokenData current = from.get(i);

            if (from.size() > i + 1) {
                
                if (current.token instanceof Value)
                    arguments.add (current);
                TokenData next = from.get(i + 1);
                
                if (next.token instanceof Comma) {
                    i++;
                }else {
                    return arguments;
                }
                
            } else {
                arguments.add(current);
                return arguments;
            }
        }
        
        return arguments;
    }

    // sqrt (pow (10, 2) + pow (10, 2))
    
    @Override
    public EvaluateResult evaluate(ArrayList<TokenData> allTokens, int thisIndex) {
        
        ArrayList<TokenData> arguments = getArguments (allTokens, thisIndex);
        if (arguments.size () != argCount) throw new IllegalArgumentException ("Invalid argument amount for " + toString () + ". Expected " + argCount + ", got " + arguments.size () + ".");
        
        int[] spendTokens = new int[arguments.size () * 2];
        if (arguments.isEmpty())
            spendTokens = new int[1];
        
        double[] values = new double[arguments.size ()];
        spendTokens[0] = thisIndex;
        
        for (int i = 0; i < arguments.size (); i++) {
            spendTokens[i * 2 + 1] = thisIndex + i * 2 + 1; // Values
            
            if (i != arguments.size() - 1) {
                spendTokens[i * 2 + 2] = thisIndex + i * 2 + 2; // Commas
            }
            
            values[i] = ((Value) arguments.get(i).token).getValue();
        }
        
        return new EvaluateResult(function.evaluate(values), spendTokens);
    }

    @Override
    public String getIdentifier() {
        return name;
    }

    @Override
    public ITokenable getNew() {
        return new Function(name, precedence, argCount, function);
    }
    
    @Override
    public String toString () {
        return getIdentifier ();
    }

}

File: C#/MathematicalExpressionParser/Tokens/Function.cs
using Lomztein.MathematicalExpressionParser.Evaluation;
using System;
using System.Collections.Generic;

namespace Lomztein.MathematicalExpressionParser.Token {

    public class Function : IToken, ITokenable, IEvaluable {

        public Function(String name, int precedence, int argCount, Func<double[], double> function) {
            Name = name;
            Precedence = precedence;
            Func = function;
            ArgCount = argCount; ;
        }

        public string Name { get; set; }
        public string Identifier { get => Name; set => Name = value; }
        public int Precedence { get; set; }
        public Func<double[], double> Func { get; set; }
        public int ArgCount { get; set; }

        private List<Tuple<IToken, int>> GetArguments(List<Tuple<IToken, int>> from, int startIndex) {

            var arguments = new List<Tuple<IToken, int>> ();
            for (int i = startIndex + 1; i < from.Count; i++) {

                var current = from[i];

                if (from.Count > i + 1) {

                    if (current.Item1 is Value)
                        arguments.Add (current);
                    var next = from[i + 1];

                    if (next.Item1 is Comma) {
                        i++;
                    } else {
                        return arguments;
                    }

                } else {
                    arguments.Add (current);
                    return arguments;
                }
            }

            return arguments;
        }

        public (double result, int[] spendIndicies) Evaluate(List<Tuple<IToken, int>> allTokens, int thisIndex) {

            var arguments = GetArguments (allTokens, thisIndex);
            if (arguments.Count != ArgCount) throw new InvalidOperationException ("Invalid argument amount for " + ToString () + ". Expected " + ArgCount + ", got " + arguments.Count + ".");

            int[] spendTokens = new int[arguments.Count * 2];
            if (arguments.Count == 0)
                spendTokens = new int[1];

            double[] values = new double[arguments.Count];
            spendTokens[0] = thisIndex;

            for (int i = 0; i < arguments.Count; i++) {
                spendTokens[i * 2 + 1] = thisIndex + i * 2 + 1; // Values

                if (i != arguments.Count - 1) {
                    spendTokens[i * 2 + 2] = thisIndex + i * 2 + 2; // Commas
                }

                values[i] = ((Value)arguments[i].Item1).Number;
            }

            return (Func (values), spendTokens);
        }

        public ITokenable GetNew() {
            return new Function (Name, Precedence, ArgCount, Func);
        }

        public override string ToString() => Identifier;

    }
}
